### PR TITLE
Add modules for deterministic total ordering derivation from partial ord

### PR DESCRIPTION
**Pull request by ValarDragon**
Component of #1329 

## What is the purpose of the change

The intention of this logic is to allow us to specify partial orderings when orderings are needed (e.g. beginblock, endblock, upgrade orders),and then have an automatic derivation of the relevant total ordering.

The problem faced right now is the important relations are obscured in app.go at the moment, and its a lot of tedious logic to verify, for an area that can have subtle errors slip through. The aim of this change is to remedy this.

## Brief change log

- Does not alter the state machine, I suggest we do that in a second PR where we can verify the important parts of the ordering carefully
- Adds a DAG data structure, with deterministic topological sort. Uses the DAG + topological sort to derive a total ordering. https://en.wikipedia.org/wiki/Topological_sorting#Relation_to_partial_orders

## Testing and Verifying

This change added tests and can be verified as follows:

  - dag_test.go provides unit tests for AddFirst and TopologicalSort
  - partialord_test.go demonstrates every exposed API, what it will look for the intended use case in app module ordering, and that the sorted result obeys intended properties.
  - If you want to verify the logic, the dag uses adjacency lists to store edges, and sorts using [Kahn's algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm) (The algo is straightforward). The Dag is not intended to be externally used, its solely for the partial ord package, hence the internal package syntax employed.

I mainly wrote this for fun, its not really a priority if people don't have bandwidth to review. If you have a couple minutes, would appreciate feedback on if others agree `partialord_test.go` is a better API than what exists right now (https://github.com/osmosis-labs/osmosis/blob/ed1011699856fed0cab14d2442604b4ea9df6526/app/modules.go#L225-L254) and if we should switch to it, pending later logic agreements.

This PR intentionally does not change app.go, and just shows the API via a test, so that we can do a moreful check of all orderings independently of logic / agreeing on the approach.

## Documentation

  - Does this introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? Both modules are intended to be documented by their module.go's and function godocs.

### CODE REVIEW

File: osmoutils/partialord/partialord_test.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	moduleNames := []string{"auth", "authz", "bank", "capabilities",
+		"staking", "distribution", "epochs", "mint", "upgrades", "wasm", "ibc",
+		"ibctransfers", "bech32ibc"}
+	beginBlockOrd := partialord.NewPartialOrdering(moduleNames)
+	beginBlockOrd.FirstElements("upgrades", "epochs", "capabilities")
+	beginBlockOrd.After("ibctransfers", "ibc")
+	beginBlockOrd.After("bech32ibc", "ibctransfers")
+	beginBlockOrd.Before("mint", "distribution")
+	// This is purely just to test last functionality, doesn't make sense in context
+	beginBlockOrd.LastElements("auth", "authz", "wasm")
```
Comment (ValarDragon) [inline osmoutils/partialord/partialord_test.go:23]: This should be what the app.go logic ends up looking like. 

I think in practice, we will just use `Before` syntax, and not mix `After` and `Before`

**Comment by ValarDragon** [comment]
I did a brief survey of other dag implementations before implementing this one, and I couldn't find any that followed either of:
* Being maintained & publicly exposed
* Being deterministic

Folks like hashicorp maintain an unexposed dag internally for dependency tracking things, but they don't have the same determinism requirements we have in the SDK. They put it in an internal package, so we couldn't use it even if we wanted. Seemed excessive to copy that rather than just adding the short file here imo, especially given how simple it is for this use case, and the ease of custom helpers like AddFirstElements and AddLastElements.

File: osmoutils/partialord/internal/dag/dag.go [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+	nodeIds, err := dag.namesToIds(nodes)
+	if err != nil {
+		return err
+	}
+
+	return dag.addLast(nodeIds)
+}
+
+func (dag *Dag) addLast(nodes []int) error {
+	nodeMap := map[int]bool{}
```
Comment (p0mvn) [inline osmoutils/partialord/internal/dag/dag.go:170]: nit:
```suggestion
	nodeMap := map[int]struct{}
```

**Review by p0mvn — CHANGES_REQUESTED** [review]
I think this is an awesome design, was really interesting to review this. Great work! I really like how much more readable and less prone to user input errors the new API will make it in `app.go`

Most of my comments are readability-related nits. However, I think that we should add more tests:
- Test for `AddLast` in `dag_test.to` 
- Tests in `partialord_test.go` related to the sequence of calls of the API
   * Aside from helping to spot bugs, if any, these tests should be helpful to understand the usage of the API

File: osmoutils/partialord/internal/dag/dag.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
+package dag
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:1]: `internal` is a special folder in golang paths, that prevents external packages from importing it. https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface

File: osmoutils/partialord/internal/dag/dag.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+import (
+	"fmt"
+	"sort"
+)
+
+// Dag struct maintains a directed acyclic graph, using adjacency lists to track edges.
+type Dag struct {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:9]: The goal of it being internal is noone externally is allowed to use it!

File: osmoutils/partialord/internal/dag/dag.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+	// there is a directed edge from u -> v, if directedEdgeList[u][v] = 1
+	// there is a directed edge from v -> u, if directedEdgeList[u][v] = -1
+	directedEdgeList []map[int]int8
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:12]: It can, but do you think that would be simpler? I found the negative relation simpler for groking direction

(Don't need to optimize for performance here, at relevant sizes its negligible)

File: osmoutils/partialord/internal/dag/dag.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+	return Dag{
+		directedEdgeList: directedEdgeList,
+		nodeNames:        nodeNames,
+		idToNodeNames:    idToNodeNames,
+	}
+}
+
+// Copy returns a new dag struct that is a copy of the original dag.
+// Edges can be mutated in the copy, without altering the original.
+func (dag Dag) Copy() Dag {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:38]: I generally think its bad API design within golang that Copy/Clone isn't more normally exposed for everything. Happy to make it internal though!

File: osmoutils/partialord/internal/dag/dag.go (hunk 2 above)
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:170]: Is this more idiomatic for existence?

File: osmoutils/partialord/internal/dag/dag.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+		sortedIDs = append(sortedIDs, n)
+		nEdgeList := G.directedEdgeList[n]
+
+		// normally we'd do map iteration, but because we need cross-machine determinism,
+		// we gather all the nodes M for which there is an edge n->m, sort that list,
+		// and then iterate over it.
+		nodesM := make([]int, 0, len(nEdgeList))
+		for m, direction := range nEdgeList {
+			if direction != 1 {
+				panic("dag: correctness error")
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:282]: yeah, n should only be selected when it has no incoming edges, and only outgoing edges.

So this edge being inbound should never happen by the rest of the logic being valid

File: osmoutils/partialord/partialord_test.go (hunk 1 above)
Comment (ValarDragon) [reply osmoutils/partialord/partialord_test.go:23]: I don't think so, what makes sense depends on the context of a given application

**Comment by ValarDragon** [comment]
Thanks for the detailed review!

File: osmoutils/partialord/internal/dag/dag.go [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+// Returns a Topological Sort of the DAG, using Kahn's algorithm.
+// https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+func (dag Dag) TopologicalSort() []string {
+	// G is the mutable graph we work on, which we remove edges from.
+	G := dag.Copy()
+	// L in pseudocode
+	sortedIDs := []int{}
+	topLevelNodes := dag.topologicalTopLevelNodes()
+
+	// while len(topLevelNodes) != 0
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:264]: My thinking was to make it obvious to anyone reading the wikipedia for line to line logic comparison

File: osmoutils/partialord/internal/dag/dag.go [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+func (dag *Dag) AddFirstElements(nodes ...string) error {
+	nodeIds, err := dag.namesToIds(nodes)
+	if err != nil {
+		return err
+	}
+
+	return dag.addFirst(nodeIds)
+}
+
+func (dag *Dag) addFirst(nodes []int) error {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:124]: Not sure this helps readability of the function logic. https://dave.cheney.net/2019/01/29/you-shouldnt-name-your-variables-after-their-types-for-the-same-reason-you-wouldnt-name-your-pets-dog-or-cat

File: osmoutils/partialord/internal/dag/dag.go [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+	nodeMap := map[int]bool{}
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:125]: I don't think this we should think about this sort of space saving in non-sensitive logic. (It has no impact on overall system performance, and takes cognitive overhead)

I would prefer not doing this unless its more idiomatic

File: osmoutils/partialord/partialord.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	err := ord.dag.AddEdge(A, B)
+	handleDagErr(err)
+}
+
+// Sets elems to be the first elements in the ordering.
+// So if were making an ordering over {A, B, C, D, E}, and elems provided is {D, B, A}
+// then we are guaranteed that the total ordering will begin with {D, B, A}
+func (ord *PartialOrdering) FirstElements(elems ...string) {
+	if ord.firstSealed {
+		panic("FirstElements has already been called")
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/partialord.go:52]: This follows the sealed idiom we use throughout the SDK

https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/options.go#L80-L82

File: osmoutils/partialord/module.go [hunk 12]
```diff
@@ -0,0 +1,2 @@
+// package partialord allows one to define partial orderings, and derive a total ordering
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/module.go:1]: Talking about the concrete algorithm used at the godoc level seems like an abstraction violation though

File: osmoutils/partialord/partialord_test.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/osmosis-labs/osmosis/v7/osmoutils/partialord"
+)
+
+func TestAPI(t *testing.T) {
+	// begin block use case, we have a dozen modules, but only care about a couple orders.
+	// In practice this will be gotten from some API, e.g. app.AllModuleNames()
+	moduleNames := []string{"auth", "authz", "bank", "capabilities",
+		"staking", "distribution", "epochs", "mint", "upgrades", "wasm", "ibc",
+		"ibctransfers", "bech32ibc"}
+	beginBlockOrd := partialord.NewPartialOrdering(moduleNames)
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/partialord_test.go:17]: Yeah sure we can add tests for it! Its supported, just FirstElements/LastElements will override it.

File: osmoutils/partialord/internal/dag/dag.go [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+func (dag Dag) Copy() Dag {
+	directedEdgeList := make([]map[int]int8, len(dag.nodeNameToId))
+	for i := 0; i < len(dag.nodeNameToId); i++ {
+		originalEdgeList := dag.directedEdgeList[i]
+		directedEdgeList[i] = make(map[int]int8, len(originalEdgeList))
+		for k, v := range originalEdgeList {
+			directedEdgeList[i][k] = v
+		}
+	}
+	// we re-use nodeNameToId and idToNodeNames as these are fixed at dag creation.
```
Comment (mattverse) [inline osmoutils/partialord/internal/dag/dag.go:47]: What's the reason we iterate to copy Dag instead of directly using `dag.directedEdgeList`? Is that to ensure we deep copy edge list?

File: osmoutils/partialord/internal/dag/dag.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattverse) [reply osmoutils/partialord/internal/dag/dag.go:12]: For the purpose of groking direction, agreed that negative relation would be more accurate, but considering code readability I'd agree with @p0mvn  to change this to boolean!

File: osmoutils/partialord/internal/dag/dag.go [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+	}
+	dag.directedEdgeList[u] = map[int]int8{}
+}
+
+// deleteEdge deletes edges between u and v.
+func (dag *Dag) deleteEdge(u, v int) {
+	delete(dag.directedEdgeList[u], v)
+	delete(dag.directedEdgeList[v], u)
+}
+
```
Comment (mattverse) [inline osmoutils/partialord/internal/dag/dag.go:100]:
```suggestion

// AddEdge checks if either edge between u and v exists and adds a directed edge from u -> v
```

**Review by mattverse — APPROVED** [review]
LGTM! This is an awesome implementation, and I love how we could existing problems of manually listing orders of Enblockers using dag and total ordering!

General question I have would be what were the possible places in this implementation that could have caused non determinism and how do we ensure "deterministic ordering"?

File: osmoutils/partialord/internal/dag/dag.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:12]: How does code readability improve with bools, if the direction is simpler with negation?

File: osmoutils/partialord/internal/dag/dag.go (hunk 14 above)
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:47]: Yeah, the topological sort algorithm used basically removes edges from every node marked as 'done', So we have to remove these edges from a copy.

More space efficient things can technically be done, but should never be important at the relevant sizes here.

File: osmoutils/partialord/partialord_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/partialord_test.go:17]: Good call on the tests, adding that test caught some incorrect behavior ;)

File: osmoutils/partialord/partialord_test.go (hunk 1 above)
Comment (p0mvn) [reply osmoutils/partialord/partialord_test.go:23]: If my understanding of the inverse relationship between `Before` and `After` is right (i.e. `beginBlockOrd.After("ibctransfers", "ibc")` is the same as `beginBlockOrd.Before("ibc", "ibctransfers")`)

I'm still not quite sure, why do we need to expose both? In what context do we need both if we can substitute one for the other by switching the order of arguments?

File: osmoutils/partialord/internal/dag/dag.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:12]: Although I had read the comment, I didn't build that link in my mind between the direction and the integer sign while reviewing. However, after you pointed it out, it makes perfect sense. It would probably be easier for me to review if I originally understood that link

I think representing binary values as booleans might feel more readable because if a developer reads one of the `Dag` methods and sees `directedEdgeList` map values used as `int8` before seeing this comment, it might confuse them at first. The reason for that is that they have to read the comment to understand that out of 2^8 possible values, only 2 are used.

It seems that both options have different pros and cons. @ValarDragon it's up to you which one to keep

File: osmoutils/partialord/internal/dag/dag.go (hunk 2 above)
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:182]: Since there are no sets in Go, this idiom is commonly used in place of sets where it avoids spending an unnecessary byte for `bool` in each key/value pair:
- https://www.davidkaya.com/sets-in-golang/
- https://dave.cheney.net/2014/03/25/the-empty-struct

File: osmoutils/partialord/internal/dag/dag.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:9]: `internal` package makes sense but here I was referring to renaming `Dag` -> `DAG`

File: osmoutils/partialord/internal/dag/dag.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:38]: I'm generally for avoiding exporting the API that is unused outside of the exporting package but I understand your view as well. Feel free to resolve it as is :+1:

File: osmoutils/partialord/module.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/module.go:1]: True, I agree. I saw you added the link within the PartialOrdering struct definition. That makes more sense, thanks for adding it

File: osmoutils/partialord/partialord.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/partialord.go:56]: While I'm not feeling strongly about this, just FYI that someone suggested changing this idiom in one of my reviews in the SDK: https://github.com/cosmos/cosmos-sdk/issues/11600

I think that their reasoning for that here is pretty solid: https://github.com/cosmos/cosmos-sdk/pull/11496#discussion_r847363424

The second point from that argument applies most here: `if the app is sealed then the correct behavior is obvious: just don't do anything and return`. Similarly, if `ord` is `firstSealed`, doing nothing makes most sense

File: osmoutils/partialord/internal/dag/dag.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:135]: Replied with more info in another comment: https://github.com/osmosis-labs/osmosis/pull/1335#discussion_r857155666

tldr; this is idiomatic with sets in Go

File: osmoutils/partialord/internal/dag/dag.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:134]: I think your reply was in respone to [another](https://github.com/osmosis-labs/osmosis/pull/1335/files/b53df15ffcbf38e781df4f13256b1e8e73a8f063#r857060889) comment?

I suggested the name `nodeIds` here because nodes may have a:
1. name
2. id

When I was reading this, I got stumbled because `nodes` has two meanings. Although it's implied that the type of `nodes` refers to option 2, it got me thinking whether `nodes` may I have a third integer meaning

File: osmoutils/partialord/internal/dag/dag.go [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+	"fmt"
+	"sort"
+)
+
+// Dag struct maintains a directed acyclic graph, using adjacency lists to track edges.
+type Dag struct {
+	// there is a directed edge from u -> v, if directedEdgeList[u][v] = 1
+	// there is a directed edge from v -> u, if directedEdgeList[u][v] = -1
+	directedEdgeList []map[int]int8
+	nodeNames        map[string]int
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go]: @ValarDragon I think you replied to this in another thread with:

> Not sure this helps readability of the function logic. https://dave.cheney.net/2019/01/29/you-shouldnt-name-your-variables-after-their-types-for-the-same-reason-you-wouldnt-name-your-pets-dog-or-cat

I agree with the article. However, I suggest changing this to something that explains the purpose of the value e.g. `nodeNamesToIds`. The article refers to the types whereas this suggestion refers to the logical meaning of the contents of the map.

At first, it wasn't immediately apparent to me what the value of this map represents

[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go]: It already was renamed to `nodeNameToIds`! (Also why I didn't reply that to this comment :) )

File: osmoutils/partialord/internal/dag/dag.go (hunk 2 above)
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:182]: I don't think this is a nit that matters. (Nor do I think either article leads one to the conclusion around whats idiomatic in golang. The Dave Cheney blog doesn't advocate for this in a map. Go Nuts developer threads don't have clear consensus either.

In a larger scheme, this is a micro-optimization, that has no impact on system performance. Do you find this a readability hurdle?

File: osmoutils/partialord/partialord.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/partialord.go:56]: I think peter was wrong there, for a similar reason that it doesn't make sense here.

That second point is especially wrong in the context of the SDK, as the entire SDK design philosophy is not to fail silently.

File: osmoutils/partialord/partialord_test.go (hunk 1 above)
Comment (ValarDragon) [reply osmoutils/partialord/partialord_test.go:23]: Similar reason to why int libraries expose `>` and `<=`

File: osmoutils/partialord/internal/dag/dag.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go:12]: Going to keep negatives

File: osmoutils/partialord/internal/dag/dag.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/partialord/internal/dag/dag.go]: I tried the change before posting that, and found the code / comments concretely harder to parse personally

File: osmoutils/partialord/internal/dag/dag.go (hunk 2 above)
Comment (p0mvn) [reply osmoutils/partialord/internal/dag/dag.go:182]: I generally try to avoid leaving out micro-optimizations if there is a language construct that allows mitigating the overhead. As the repository evolves, the use patterns change and the overhead compounds, such things might become problematic.

I'm also used to seeing this pattern across codebases so it doesn't feel unreadable to me. 

However, since all my points are a matter of personal preference and trying to optimize something negligible, I'm okay with keeping everything as is.

Thanks for being thorough in addressing and discussing this comment :+1:

**Review by p0mvn — APPROVED** [review]
LGTM
